File: Apps4Bro.Android/old/java/AdManager.java
package net.runserver.apps4bro;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URL;
import java.net.URLConnection;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences.Editor;
import android.util.Log;

public class AdManager
{
	private final static String TAG = "AdManager";
    private final static String App4BroTag = "app4bro";

	private Context m_context;

	private final Map<String, String> m_keys;
	private final Map<String, AdNetworkHandler> m_adNetworks;
	
	private AdWrapper[] m_adWrappers;
	private int m_currentWrapper;
	private boolean m_adShown;
    private String m_appId;

	private static AtomicBoolean s_asyncTaskStarted = new AtomicBoolean(false);

	public boolean isAdShown()
	{
		return m_adShown;
	}
	
	public Context getContext()
	{
		return m_context;
	}
	
	public AdManager(String[] keys)
	{
		m_keys = new HashMap<String, String>();

		for (int i = 0; i < keys.length / 2; i++)
			m_keys.put(keys[i * 2].toLowerCase(), keys[i * 2 + 1]);
		
        m_adNetworks = new HashMap<String, AdNetworkHandler>();
        
		m_appId = m_keys.get(App4BroTag);
        if (m_appId == null)
        {
        	Log.e(TAG, "No App4Bro key specified for AdManager!");
        	return;
        }

        ////////
        registerAdNetwork(new AdMobNetwork(this));
        registerAdNetwork(new AppodealNetwork(this));
       // registerAdNetwork(new AppodealVideoNetwork(this));
        registerAdNetwork(new AppLovinNetwork(this));
        registerAdNetwork(new AdCashNetwork(this));
	}

	public boolean isAdTimeoutCompleted(Context context)
	{
		long now = new Date().getTime();
		long next = context.getSharedPreferences(context.getPackageName() + "_preferences", Context.MODE_PRIVATE).getLong("next_ad", -1);

		if (next == -1)
		{
			next = now + 1; // next launch
			Editor editor = context.getSharedPreferences(context.getPackageName() + "_preferences", Context.MODE_PRIVATE).edit();
			editor.putLong("next_ad", next);
			editor.commit();
		}
		return next <= now;
	}
	
	private void setAdTimeout()
	{
		Editor editor = m_context.getSharedPreferences(m_context.getPackageName() + "_preferences", Context.MODE_PRIVATE).edit();
		editor.putLong("next_ad", new Date().getTime() + 180000); // 3 minutes
		editor.commit();	
	}
	
    public void registerAdNetwork(AdNetworkHandler handler)
    {
        m_adNetworks.put(handler.getNetwork().toLowerCase(), handler);
    }

	
	public void init(Context context)
	{
		m_context = context;
		
		/*
		
		AsyncTask<Void, Void, Void> task = new AsyncTask<Void, Void, Void>() {
			@Override
			protected Void doInBackground(Void... params)
			{
				initAsync();
				return null;
			}
			
			@Override
			protected void onPostExecute(Void result)
			{
				nextAd();
			}
		};
		task.execute();*/
		
		if (s_asyncTaskStarted.get() == false)
		{
			Thread thread = new Thread(new Runnable(){
				  @Override
				  public void run(){
				    initAsync();
				  }
				});
				thread.start();
			}
	}
	
	private void initAsync()
	{
		s_asyncTaskStarted.set(true);

		try
		{
			URLConnection conn = new URL("http://runserver.net/app4bro/ad.php?app=" + m_appId).openConnection();
			conn.setConnectTimeout(5000);
			conn.setReadTimeout(5000);
			
			String data = loadStreamText(conn.getInputStream());
			if (data != null && data.length() > 0)
				parseAdData(data);
			else
				Log.w(TAG, "Got empty response");
			
		} catch (Exception ex)
		{
			Log.w(TAG, "Network error: " + ex);
			ex.printStackTrace();
		}

		if (m_adWrappers == null)
		{
            Log.d(TAG, "Registering fallback ads");

            StringBuilder builder = new StringBuilder();
            
            int count = 0;
            for(Map.Entry<String,String> pair : m_keys.entrySet())
            {
                count++;
                builder.append(pair.getKey());
                builder.append(':');
                builder.append(pair.getKey());
                builder.append("def");
                builder.append(count);
                builder.append('|');
                builder.append(pair.getValue());
                builder.append('|');
            }
            
            parseAdData(builder.toString());
			
			if (m_adWrappers == null || m_adWrappers.length == 0)
			{
				Log.w(TAG, "No ad networks registered!");
				return;
			}
		}

		m_currentWrapper = -1;
		
		((Activity)m_context).runOnUiThread(new Runnable()
		{
			  @Override
			  public void run(){
			    showAd();
			  }
			});
				
		
		s_asyncTaskStarted.set(false);
	}
	
	private void parseAdData(String data)
	{
		try
		{
			String[] adData = data.split("\\|");
	
			if (adData.length > 1)
			{
				List<AdWrapper> wrappers = new ArrayList<AdWrapper>(adData.length / 2);
	
				for (int i = 0; i < adData.length; i += 2)
				{
					String network = adData[i].toLowerCase();
	                String networkId;
	
	                if (network.contains(":"))
	                {
	                    String [] nsplit = network.split(":");
	                    network = nsplit[0];
	                    networkId = nsplit[1];
	                }
	                else
	                    networkId = network + " " + (wrappers.size() + 1);
	
	                AdNetworkHandler networkHandler = m_adNetworks.get(network);
	
					if (networkHandler != null)
					{
						wrappers.add(new AdWrapper(networkHandler, adData[i + 1], networkId));
						Log.d(TAG, "Registered ad network " + network + "[" + adData[i+1] + "]");
					} else
						Log.w(TAG, "Failed to register ad network " + network);
				}
	
				if (wrappers.size() == 0)
				{
					Log.w(TAG, "No ad networks registered!");
				} else
				{
					m_adWrappers = wrappers.toArray(new AdWrapper[wrappers.size()]);
					Log.d(TAG, "Registered " + m_adWrappers.length + " ad networks");
				}
			}
		}
		catch(Exception ex)
		{
			Log.e(TAG, "Error parsing data string: " + data);
			ex.printStackTrace();
		}
	}
	
	public void showAd()
	{
        if (m_adNetworks == null)
        {
            Log.w(TAG, "showAd called before initialization!");
            return;
        }
        
		m_currentWrapper++;
		
		Log.w(TAG, "Going to use wrapper #" + m_currentWrapper);
		
		if (m_currentWrapper >= m_adWrappers.length)
		{
			Log.w(TAG, "No ad networks left for showAd()!");
			m_currentWrapper = -1;
			return;
		}

		AdWrapper wrapper = null;
		try
		{
            wrapper = m_adWrappers[m_currentWrapper];
            wrapper.CallCount++;
            
            wrapper.getNetworkHandler().setId(wrapper.getId());
			
			if (!wrapper.getNetworkHandler().show(wrapper))
				adError(wrapper, "Wrapper returned false");
			
		} catch (Exception ex)
		{
			ex.printStackTrace();
            adError(wrapper, ex.getMessage());			
		}
	}
	
    public void hideAd()
    {
        if (m_adNetworks == null)
        {
            Log.w(TAG, "hideAd called before initialization!");
            return;
        }
        
        m_adShown = false;
        
        AdWrapper wrapper = null;
        try
        {
            wrapper = m_adWrappers[m_currentWrapper];
            wrapper.getNetworkHandler().hide();
            Log.d(TAG, "Called hide ad for network "  + wrapper.getNetworkHandler().getNetwork());
        }
        catch (Exception ex)
        {
        	ex.printStackTrace();
        }
    }
	
    /*internal*/ void adError(Object data, String error)
    {
        String networkName = "Unknown network";
        try
        {
            AdWrapper wrapper = (AdWrapper)data;

            if (wrapper != null)
            {
                networkName = wrapper.getName();
                wrapper.LastError = error;
                wrapper.FailCount++;
                Log.w(TAG, String.format("Failed to display ad from %s [%d/%d/%d/%d], error %s", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount, error));
                //m_reportManager.ReportEvent("AD_ERROR", string.Format("{0} [{1}/{2}/{3}/{4}] '{5}'", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount, error));
            }
            else
            {
                Log.w(TAG, "Failed to display ad from " + networkName + ", error " + error);
                //m_reportManager.ReportEvent("AD_ERROR", string.Format("{0} '{1}'", networkName, error));
            }

        }
        finally
        {
            //if (m_context.OnFailedAd(this, networkName, error))
        	if (m_currentWrapper == - 1 || m_adWrappers[m_currentWrapper] == data)
                showAd();
        }
    }
	
	/*internal*/ void adLoaded(Object data, Object view)
	{
        String networkName = "Unknown network";

        try
        {
        	setAdTimeout();
		
	        AdWrapper wrapper = (AdWrapper)data;
	        
	        m_adShown = true;
	
	        if (wrapper != null)
	        {
	            networkName = wrapper.getName();
	            wrapper.SuccessCount++;
	            Log.w(TAG, String.format("Displayed ad from %s [%d/%d/%d/%d]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount));
//	            m_reportManager.ReportEvent("AD_SHOW", string.Format("{0} [{1}/{2}/{3}/{4}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount));
	        }
	        else
	        {
	            Log.w(TAG, "Displayed ad from " + networkName);
//	            m_reportManager.ReportEvent("AD_SHOW", "");
	        }
	        
	    }
	    finally
	    {
	        //m_context.OnShownAd(this, networkName, view);
	    }
	}

	/*internal*/ void adClick(Object data)
	{
        String networkName = "Unknown network";

        try
        {
	        AdWrapper wrapper = (AdWrapper)data;
	        
	        m_adShown = false;
	
	        if (wrapper != null)
	        {
	            networkName = wrapper.getName();
	            wrapper.ClickCount++;
	            Log.w(TAG, String.format("Clicked ad from %s [%d/%d/%d/%d]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount));
//                m_reportManager.ReportEvent("AD_CLICK", string.Format("{0} [{1}/{2}/{3}/{4}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount));
	        }
	        else
	        {
	            Log.w(TAG, "Displayed ad from " + networkName);
//	            m_reportManager.ReportEvent("AD_SHOW", "");
	        }
	        
	    }
	    finally
	    {
//            m_context.OnClickedAd(this, networkName);
	    }
	}
	
	private static String loadStreamText(InputStream stream)
	{
		try
		{
			BufferedReader reader = new BufferedReader(new InputStreamReader(stream, "UTF-8"), 0x8000);

			StringBuilder result = new StringBuilder();
			String line = null;

			while ((line = reader.readLine()) != null)
			{
				result.append(line);
				// Log.d("Test", "Line: " + line);
			}

			return result.toString();
		} catch (Exception ex)
		{
			ex.printStackTrace();

		}
		return "";
	}
	
}

File: Apps4Bro.iOS/AdManager.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net;
using System.IO;
using System.Text;
using Apps4Bro.Networks;
using UIKit;

#if __IOS__
using Foundation;
#endif

namespace Apps4Bro
{
    public interface AdNetworkHandler
    {
        bool Show(object data);

        void Hide();

        void Display();

        void SetId(string id);

        string Network { get; }
    }

    public class AdManager
    {
        private readonly static string App4BroTag = "app4bro";

        private readonly Dictionary<string, AdNetworkHandler> m_adNetworks;
        private readonly Dictionary<string, string> m_keys;

        private readonly ReportManager m_reportManager;
#if __IOS__
        private readonly NSOperationQueue m_operationQueue = new NSOperationQueue();
#endif

        private AdWrapper[] m_adWrappers;
        private volatile AdContextDelegate m_context;
        private int m_currentWrapper;
        private bool m_adShown;
        private bool m_adLoaded;
        private bool m_inited;
        private string m_appId;

        public bool IsAdShown
        {
            get { return m_adShown; }
        }

        public bool IsAdLoaded
        {
            get { return m_adLoaded; }
        }

        public bool IsInited
        {
            get { return m_adNetworks != null && m_adWrappers != null; }
        }

        public AdContextDelegate Context
        {
            get { return m_context; }
        }

        public AdManager(params string[] keys)
        {
            m_keys = new Dictionary<string, string>();

            for (int i = 0; i < keys.Length; i += 2)
                m_keys.Add(keys[i].ToLower(), keys[i + 1]);

            if (!m_keys.TryGetValue(App4BroTag, out m_appId))
                throw new ApplicationException("No App4Bro key specified for AdManager!");

            m_reportManager = new ReportManager(m_appId);

            m_reportManager.ReportEvent("AD_START", Apps4BroSDK.Version.ToString());

            m_adNetworks = new Dictionary<string, AdNetworkHandler>();

            // Registering built-in SDKs
#if __IOS__
            // Interstitials
#if USE_ADMOB
            RegisterAdNetwork(new AdMobNetwork(this));
#endif
#if USE_ADCASH
            RegisterAdNetwork(new AdCashNetwork(this));
#endif
#if USE_IAD
            RegisterAdNetwork(new iAdNetwork(this));
#endif

            if (UIKit.UIDevice.CurrentDevice.CheckSystemVersion(7, 0))
            {
#if USE_APPODEAL
                RegisterAdNetwork(new AppodealNetwork(this));
#endif
#if USE_INNERACTIVE
                RegisterAdNetwork(new InneractiveNetwork(this));
#endif
            }

            // Banners
#if USE_ADMOB
            RegisterAdNetwork(new AdMobBannerNetwork(this));
#endif
#if USE_IAD
            RegisterAdNetwork(new iAdBannerNetwork(this));
#endif


            if (UIKit.UIDevice.CurrentDevice.CheckSystemVersion(7, 0))
            {
#if USE_INNERACTIVE
                RegisterAdNetwork(new InneractiveBanner(this));
#endif
#if USE_ADTOAPP
                RegisterAdNetwork(new AdToAppNetwork(this));
#endif
            }

#elif __ANDROID__
            RegisterAdNetwork(new AdMobNetwork(this));
#if USE_ADCASH
            RegisterAdNetwork(new AdCashNetwork(this));
#endif
#endif
            RegisterAdNetwork(new DummyNetwork(this));
            RegisterAdNetwork(new HouseNetwork(this));

            PreInit();
        }

        public void RegisterAdNetwork(AdNetworkHandler handler)
        {
            m_adNetworks[handler.Network.ToLower()] = handler;
        }

        private string FormatRequest()
        {
            return string.Format(Apps4BroSDK.AdManagerUrl,
                Apps4BroSDK.AdvertisingId,
                m_appId
            );
        }

        private void PreInit()
        {

#if __IOS__
            NSUrl nsurl = NSUrl.FromString(FormatRequest());

            NSMutableUrlRequest request = new NSMutableUrlRequest(nsurl);
            request.TimeoutInterval = 5.0;

            NSUrlConnection.SendAsynchronousRequest(request, m_operationQueue, (response, ndata, error) =>
                UIApplication.SharedApplication.BeginInvokeOnMainThread(() =>
                    DoInit(ndata != null ? ndata.ToString() : null, error != null ? (HttpStatusCode)(int)error.Code : HttpStatusCode.OK)
                )
            );
#else
            HttpWebRequest request = HttpWebRequest.Create(FormatRequest());
            request.ProtocolVersion = HttpVersion.Version20;
            request.UserAgent = "WinHTTP";
            request.ContentType = "application/text";
            request.Method = "GET";
#if !NETFX_CORE
            request.Timeout = 5000;
#endif

            request.BeginGetResponse(InitAsync, request);
#endif
        }

        public void Init(AdContextDelegate context)
        {
            if (m_context == null && m_inited)
            {
                m_context = context;
                m_context.OnInited(this);
                return;
            }

            if (IsInited)
            {
                LoadAd();
                return;
            }

            m_context = context;
        }

#if NETFX_CORE
        private void InitAsync(IAsyncResult asynchronousResult)
        {
            string adData = null;
            HttpStatusCode code = HttpStatusCode.Unused;

            try
            {

                WebRequest request = (WebRequest)asynchronousResult.AsyncState;
                HttpWebResponse response = request.EndGetResponse(asynchronousResult) as HttpWebResponse;
                code = response == null ? HttpStatusCode.Unused : response.StatusCode;

                if (response != null && code == HttpStatusCode.OK)
                {
                    using (StreamReader reader = new StreamReader(response.GetResponseStream()))
                        adData = reader.ReadToEnd();
                }

            }
            catch (Exception ex)
            {
                Debug.WriteLine("Network error: " + ex);
            }

            DoInit(adData, code);
        }
#endif
        private void DoInit(string adData, HttpStatusCode code)
        {
            try
            {
                if (adData == null || code != HttpStatusCode.OK)
                {
                    Debug.WriteLine("Error fetching data. Server returned status code: {0}", code);
                }

                if (!string.IsNullOrWhiteSpace(adData))
                {
                    ParseAdData(adData, true);
                }
                else
                    Debug.WriteLine("Got empty response");

            }
            catch (Exception ex)
            {
                Debug.WriteLine("Network error: " + ex);
            }

            if (m_adWrappers == null) // try cached data
            {
#if __IOS__

                string data = Foundation.NSUserDefaults.StandardUserDefaults.StringForKey(App4BroTag + "_" + m_appId);
                if (data != null)
                {
                    Debug.WriteLine("Registering cached ads");
                    ParseAdData(data, false);
                }

#endif
            }


            if (m_adWrappers == null)
            {
                Debug.WriteLine("Registering fallback ads");

                StringBuilder builder = new StringBuilder();

                int count = 0;
                foreach (KeyValuePair<string, string> pair in m_keys)
                {
                    count++;
                    builder.AppendFormat("{0}:{0}def{1}|{2}|", pair.Key, count, pair.Value);
                }

                ParseAdData(builder.ToString(), false);

                if (m_adWrappers == null)
                {
                    Debug.WriteLine("No ad networks registered!");
                    return;
                }
            }

            m_currentWrapper = -1;

            m_inited = true;

            if (m_context != null)
                m_context.OnInited(this);
        }

        private void ParseAdData(string data, bool save)
        {
            try
            {
                string[] adData = data.Split('|');

                if (adData.Length > 1)
                {
                    List<AdWrapper> wrappers = new List<AdWrapper>(adData.Length / 2);

                    for (int i = 0; i < adData.Length; i += 2)
                    {
                        string network = adData[i].ToLower();
                        string networkId;

                        if (network.Contains(":"))
                        {
                            string[] nsplit = network.Split(':');
                            network = nsplit[0];
                            networkId = nsplit[1];
                        }
                        else
                            networkId = network + " " + (wrappers.Count + 1);

                        AdNetworkHandler networkHandler = null;

                        if (m_adNetworks.TryGetValue(network, out networkHandler))
                        {
                            wrappers.Add(new AdWrapper(networkHandler, adData[i + 1], networkId));
                            Debug.WriteLine("Registered ad network " + network + "[" + adData[i + 1] + "]");
                        }
                        else
                            Debug.WriteLine("Failed to register ad network " + network);
                    }

                    if (wrappers.Count == 0)
                    {
                        Debug.WriteLine("No ad networks registered!");
                    }
                    else
                    {
                        m_adWrappers = wrappers.ToArray();
                        Debug.WriteLine("Registered " + m_adWrappers.Length + " ad networks");
                    }

                    if (save)
                    {
#if __IOS__
                        Foundation.NSUserDefaults.StandardUserDefaults.SetString(data, App4BroTag + "_" + m_appId);
#endif
                    }


                    m_reportManager.ReportEvent("AD_INIT", m_adWrappers.Length.ToString());
                }
            }
            catch (Exception)
            {
                Debug.WriteLine("Error parsing data string " + data);
            }
        }

        public void HideAd()
        {
            if (!IsInited)
            {
                Debug.WriteLine("HideAd called before initialization!");
                return;
            }

            if (!m_adShown)
            {
                Debug.WriteLine("HideAd called with no shown ads!");
                return;
            }

            AdWrapper wrapper = null;
            try
            {
                m_adShown = false;
                wrapper = m_adWrappers[m_currentWrapper];
                wrapper.AdNetworkHandler.Hide();
                Debug.WriteLine("Called hide ad for network " + wrapper.AdNetworkHandler.Network);
            }
            catch (Exception ex)
            {
                Debug.WriteLine("HideAd Exception: " + ex);
            }
        }

        public void LoadAd()
        {
            if (!IsInited)
            {
                throw new ApplicationException("LoadAd called before initialization!");
            }

            m_currentWrapper++;
            m_adLoaded = false;

            Debug.WriteLine("Going to use wrapper #" + m_currentWrapper);

            if (m_currentWrapper >= m_adWrappers.Length)
            {
                Debug.WriteLine("No ad networks left for LoadAd()!");
                m_currentWrapper = -1;
                return;
            }

            AdWrapper wrapper = null;
            try
            {
                wrapper = m_adWrappers[m_currentWrapper];
                wrapper.CallCount++;

                wrapper.AdNetworkHandler.SetId(wrapper.Id);

                if (!wrapper.AdNetworkHandler.Show(wrapper))
                    throw new ApplicationException("Ad wrapper " + wrapper.Name + " returned false");

            }
            catch (Exception ex)
            {
                Debug.WriteLine("LoadAd Exception: " + ex);

                AdError(wrapper, ex.Message);
            }
        }

        public void DisplayAd()
        {
            if (!IsInited)
            {
                throw new ApplicationException("LoadAd called before initialization!");
            }

            if (!m_adLoaded)
            {
                Debug.WriteLine("DisplayAd called when no ad is loaded!");
                return;
            }

            AdWrapper wrapper = null;
            try
            {
                wrapper = m_adWrappers[m_currentWrapper];
                wrapper.AdNetworkHandler.Display();
            }
            catch (Exception ex)
            {
                Debug.WriteLine("DisplayAd Exception: " + ex);

                AdError(wrapper, ex.Message);
            }
        }

        internal void AdLoaded(object data)
        {
            string networkName = "Unknown network";
            try
            {
                AdWrapper wrapper = data as AdWrapper;

                m_adLoaded = true;

                if (wrapper != null)
                {
                    networkName = wrapper.Name;
                    wrapper.SuccessCount++;
                    Debug.WriteLine("Loaded ad from {0} [{1}/{2}/{3}/{4}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount);
                    m_reportManager.ReportEvent("AD_LOAD", string.Format("{0} [{1}/{2}/{3}/{4}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount));
                }
                else
                {
                    Debug.WriteLine("Loaded ad from " + networkName);
                    m_reportManager.ReportEvent("AD_LOAD", "");
                }

            }
            finally
            {
                m_context.OnLoadedAd(this, networkName);
            }
        }

        internal void AdError(object data, string error)
        {
            string networkName = "Unknown network";
            try
            {
                AdWrapper wrapper = data as AdWrapper;

                m_adShown = false;
                m_adLoaded = false;

                if (wrapper != null)
                {
                    networkName = wrapper.Name;
                    wrapper.LastError = error;
                    wrapper.FailCount++;
                    Debug.WriteLine("Failed to load ad from {0} [{1}/{2}/{3}/{4}], error {5}", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount, error);
                    m_reportManager.ReportEvent("AD_ERROR", string.Format("{0} [{1}/{2}/{3}/{4}] '{5}'", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount, error));
                }
                else
                {
                    Debug.WriteLine("Failed to load ad from " + networkName + ", error " + error);
                    m_reportManager.ReportEvent("AD_ERROR", string.Format("{0} '{1}'", networkName, error));
                }

            }
            finally
            {
                if (m_context.OnFailedAd(this, networkName, error))
                    LoadAd();
            }
        }

        internal void AdClicked(object data)
        {
            string networkName = "Unknown network";
            try
            {
                AdWrapper wrapper = data as AdWrapper;

                if (wrapper != null)
                {
                    networkName = wrapper.Name;
                    wrapper.ClickCount++;
                    Debug.WriteLine("Clicked ad from {0} [{1}/{2}/{3}/{4}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount);
                    m_reportManager.ReportEvent("AD_CLICK", string.Format("{0} [{1}/{2}/{3}/{4}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount));
                }
                else
                {
                    Debug.WriteLine("Clicked ad from " + networkName);
                    m_reportManager.ReportEvent("AD_CLICK", networkName);
                }
            }
            finally
            {
                m_context.OnClickedAd(this, networkName);
            }
        }

        internal void AdShown(object data, object view)
        {
            string networkName = "Unknown network";
            try
            {
                AdWrapper wrapper = data as AdWrapper;

                m_adShown = true;

                if (wrapper != null)
                {
                    networkName = wrapper.Name;
                    wrapper.ShowCount++;
                    Debug.WriteLine("Shown ad from {0} [{1}/{2}/{3}/{4}/{5}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount, wrapper.ShowCount);
                    m_reportManager.ReportEvent("AD_SHOW", string.Format("{0} [{1}/{2}/{3}/{4}/{5}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount, wrapper.ShowCount));
                }
                else
                {
                    Debug.WriteLine("Shown ad from " + networkName);
                    m_reportManager.ReportEvent("AD_SHOW", networkName);
                }
            }
            finally
            {
                m_context.OnShownAd(this, networkName, view);
            }
        }
    }
}

